packages/product-json-to-csv/src/map-product-data.js

```diff
@@ -0,0 +1,195 @@
+/* @flow */
+import type {
+  ResolvedProdProj,
+  ProdWithMergedVariants,
+  SingleVarPerProduct,
+  MappedProduct,
+  Variant,
+  Category,
+  Image,
+} from 'types/product'
+
+import { flatten } from 'flat'
+import { isEmpty, reduce, pickBy } from 'lodash'
+
+export default class ProductMapping {
+  // Set flowtype annotations
+  fillAllRows: boolean
+  categoryBy: 'name' | 'key' | 'externalId' | 'namedPath'
+  lang: string
+  multiValDel: string
+
+  constructor(
+    {
+      fillAllRows = false,
+      categoryBy = 'name',
+      lang = 'en',
+      multiValueDelimiter = ';',
+    }: Object = {}
+  ) {
+    this.fillAllRows = fillAllRows
+    this.categoryBy = categoryBy
+    this.lang = lang
+    this.multiValDel = multiValueDelimiter
+  }
+
+  run(product: ResolvedProdProj) {
+    const mergedVarProduct = ProductMapping._mergeVariants(product)
+    const varWithProductInfo = ProductMapping._spreadProdOnVariants(
+      mergedVarProduct,
+      this.fillAllRows
+    )
+    const mappedProduct = varWithProductInfo.map(
+      (variant: SingleVarPerProduct) => this._mapProperties(variant)
+    )
+    const flatProducts = mappedProduct.map(flatten)
+    return flatProducts
+  }
+
+  static _mergeVariants(product: ResolvedProdProj): ProdWithMergedVariants {
+    const variant = [product.masterVariant, ...product.variants]
+    const mergedVariants = { ...product, variant }
+    delete mergedVariants.masterVariant
+    delete mergedVariants.variants
+    return mergedVariants
+  }
+
+  static _spreadProdOnVariants(
+    product: ProdWithMergedVariants,
+    fillAllRows: boolean
+  ): Array<SingleVarPerProduct> {
+    if (fillAllRows)
+      return product.variant.map((eachVariant: Variant): Object => ({
+        ...product,
+        variant: eachVariant,
+      }))
+
+    const productWithVariants: Array<Object> = product.variant.map(
+      (eachVariant: Variant): SingleVarPerProduct => ({
+        variant: eachVariant,
+      })
+    )
+    productWithVariants[0] = { ...product, ...productWithVariants[0] }
+    return productWithVariants
+  }
+
+  _mapProperties(product: SingleVarPerProduct): MappedProduct {
```

Comment:
We do so mich data wrangling here and I cant really follow why. Can we add a comment for what the input abd output shape of this is and why we need to do this?

Reply:
This part is context dependent, and will be explained in a wiki or face to face